addons/html_editor/static/src/core/dom_plugin.js

```diff
@@ -147,37 +154,34 @@ export class DomPlugin extends Plugin {
-            // <p> or <li>. The same is true for a <pre> inside a <pre>.
-            if (
-                container.childElementCount === 1 &&
-                (container.firstChild.nodeName === "P" ||
-                    container.firstChild.nodeName === "LI" ||
-                    (container.firstChild.nodeName === "PRE" && closestElement(startNode, "pre")))
-            ) {
-                const p = container.firstElementChild;
-                container.replaceChildren(...p.childNodes);
-            } else if (container.childElementCount > 1) {
-                // Grab the content of the first child block and isolate it.
-                if (
-                    isBlock(container.firstChild) &&
-                    !["TABLE", "UL", "OL"].includes(container.firstChild.nodeName)
-                ) {
-                    containerFirstChild.replaceChildren(...container.firstElementChild.childNodes);
-                    container.firstElementChild.remove();
-                }
-                // Grab the content of the last child block and isolate it.
-                if (
-                    isBlock(container.lastChild) &&
-                    !["TABLE", "UL", "OL"].includes(container.lastChild.nodeName)
-                ) {
-                    containerLastChild.replaceChildren(...container.lastElementChild.childNodes);
-                    container.lastElementChild.remove();
-                }
+        const shouldUnwrap = (node) =>
+            [...paragraphRelatedElements, "LI"].includes(node.nodeName) &&
+            block.textContent !== "" &&
+            (block.nodeName === node.nodeName ||
+                ["BLOCKQUOTE", "PRE", "DIV"].includes(block.nodeName)) &&
+            selection.anchorNode !== this.editable;
+
+        // In case the html inserted is all contained in a single root <p> or <li>
+        // tag, we take the all content of the <p> or <li> and avoid inserting the
+        // <p> or <li>.
+        if (
+            container.childElementCount === 1 &&
+            (["P", "LI"].includes(container.firstChild.nodeName) ||
+                shouldUnwrap(container.firstChild)) &&
+            selection.anchorNode !== this.editable
+        ) {
+            const p = container.firstElementChild;
+            container.replaceChildren(...p.childNodes);
+        } else if (container.childElementCount > 1) {
+            // Grab the content of the first child block and isolate it.
+            if (shouldUnwrap(container.firstChild) && !isSelectionAtStart) {
+                containerFirstChild.replaceChildren(...container.firstElementChild.childNodes);
+                container.firstElementChild.remove();
+            }
+            // Grab the content of the last child block and isolate it.
+            if (shouldUnwrap(container.lastChild) && !isSelectionAtEnd) {
+                containerLastChild.replaceChildren(...container.lastElementChild.childNodes);
+                container.lastElementChild.remove();
```

Comment:
I'm confused by this new code because it looks very different from the previous one. For example it does not seem to take into account the previous conditions of PRE inside of PRE and the previous conditions referencing TABLE, UL, OL.

Reply:
> it does not seem to take into account the previous conditions of PRE inside of PRE

`shouldUnwrap` check that if `block.nodeName === node.nodeName` which is same as PRE inside PRE.

> the previous conditions referencing TABLE, UL, OL.

We still will not unwrap `TABLE, UL, OL.` as we only unwrap paragraphRelatedElements and 'LI'.